File: frontend/src/shared/lib/theme/contrast.ts
const AA_RATIO = 4.5;
const LUMINANCE_OFFSET = 0.05;
const SRGB_CUTOFF = 0.03928;
const SRGB_LINEAR_DIVISOR = 12.92;
const SRGB_GAMMA_OFFSET = 0.055;
const SRGB_GAMMA_DIVISOR = 1.055;
const SRGB_GAMMA = 2.4;
const CHANNEL_MAX = 255;
const HEX_PAIR = 2;
const RED_WEIGHT = 0.2126;
const GREEN_WEIGHT = 0.7152;
const BLUE_WEIGHT = 0.0722;
const HEX_BODY = 6;
const HEX_PREFIX = "#";

function hexDigits(hex: string): string | null {
  if (!hex.startsWith(HEX_PREFIX)) {
    return null;
  }
  const body = hex.slice(HEX_PREFIX.length);
  if (body.length !== HEX_BODY) {
    return null;
  }
  return body;
}

function srgbChannels(hex: string): [number, number, number] | null {
  const body = hexDigits(hex);
  if (body === null) {
    return null;
  }
  const red = Number.parseInt(body.slice(0, HEX_PAIR), 16) / CHANNEL_MAX;
  const green = Number.parseInt(body.slice(HEX_PAIR, HEX_PAIR + HEX_PAIR), 16) / CHANNEL_MAX;
  const blue = Number.parseInt(body.slice(HEX_PAIR + HEX_PAIR, HEX_BODY), 16) / CHANNEL_MAX;
  if (Number.isNaN(red) || Number.isNaN(green) || Number.isNaN(blue)) {
    return null;
  }
  return [red, green, blue];
}

function linearize(channel: number): number {
  if (channel <= SRGB_CUTOFF) {
    return channel / SRGB_LINEAR_DIVISOR;
  }
  return ((channel + SRGB_GAMMA_OFFSET) / SRGB_GAMMA_DIVISOR) ** SRGB_GAMMA;
}

export function relativeLuminance(hex: string): number | null {
  const channels = srgbChannels(hex);
  if (channels === null) {
    return null;
  }
  const [red, green, blue] = channels;
  return (
    RED_WEIGHT * linearize(red) + GREEN_WEIGHT * linearize(green) + BLUE_WEIGHT * linearize(blue)
  );
}

export function contrastRatio(left: string, right: string): number | null {
  const first = relativeLuminance(left);
  const second = relativeLuminance(right);
  if (first === null || second === null) {
    return null;
  }
  const lighter = Math.max(first, second);
  const darker = Math.min(first, second);
  return (lighter + LUMINANCE_OFFSET) / (darker + LUMINANCE_OFFSET);
}

export function sufficientContrast(foreground: string, background: string): boolean {
  const ratio = contrastRatio(foreground, background);
  return ratio !== null && ratio >= AA_RATIO;
}

export function accentContrast(hex: string, white: string, nearBlack: string): string {
  const whiteRatio = contrastRatio(hex, white);
  const blackRatio = contrastRatio(hex, nearBlack);
  if (whiteRatio === null && blackRatio === null) {
    return white;
  }
  if (whiteRatio === null) {
    return nearBlack;
  }
  if (blackRatio === null) {
    return white;
  }
  return whiteRatio >= blackRatio ? white : nearBlack;
}

File: backend/app/services/theme/contrast.rb
# WCAG relative-luminance contrast (NR-48). Writes that would drop a semantic
# token below AA against its paired surface/text are rejected at the model.
module Theme
  class Contrast
    AA_RATIO = 4.5
    LUMINANCE_OFFSET = 0.05
    SRGB_CUTOFF = 0.03928
    SRGB_LINEAR_DIVISOR = 12.92
    SRGB_GAMMA_OFFSET = 0.055
    SRGB_GAMMA_DIVISOR = 1.055
    SRGB_GAMMA = 2.4
    CHANNEL_MAX = 255.0
    HEX_PAIR = 2
    RED_WEIGHT = 0.2126
    GREEN_WEIGHT = 0.7152
    BLUE_WEIGHT = 0.0722
    WHITE = "#FFFFFF"
    NEAR_BLACK = "#0E1621"

    class << self
      def sufficient?(foreground, background)
        ratio(foreground, background) >= AA_RATIO
      end

      def accent_readable?(hex)
        sufficient?(hex, WHITE) || sufficient?(hex, NEAR_BLACK)
      end

      def ratio(left, right)
        first, second = relative_luminance(left), relative_luminance(right)
        lighter, darker = [ first, second ].max, [ first, second ].min
        (lighter + LUMINANCE_OFFSET) / (darker + LUMINANCE_OFFSET)
      end

      def relative_luminance(hex)
        red, green, blue = srgb_channels(hex).map { |channel| linearize(channel) }
        (RED_WEIGHT * red) + (GREEN_WEIGHT * green) + (BLUE_WEIGHT * blue)
      end

      private

      def srgb_channels(hex)
        hex.delete("#").scan(/.{#{HEX_PAIR}}/).map { |pair| pair.to_i(16) / CHANNEL_MAX }
      end

      def linearize(channel)
        if channel <= SRGB_CUTOFF
          channel / SRGB_LINEAR_DIVISOR
        else
          ((channel + SRGB_GAMMA_OFFSET) / SRGB_GAMMA_DIVISOR)**SRGB_GAMMA
        end
      end
    end
  end
end
